Wasp Spec docs and migration guides, retire DSL/TS Config (#4157)
wasp-lang/wasp: web/docs/general/spec.md

---
title: Wasp Spec (main.wasp.ts)
---

You define and configure the high level of your app (pages, routes, queries, actions, auth, ...) in a `main.wasp.ts` file in the root of your project. We call this file the **Wasp Spec**.

You write the Wasp Spec in TypeScript, so you get out-of-the-box support in all editors, full type checking, and the flexibility of a real programming language while configuring your app.

:::info Coming from an older version of Wasp?
The Wasp Spec replaces two older ways of configuring a Wasp app:

- The **Wasp DSL** (`main.wasp`). See [Migrating from the Wasp DSL](../guides/legacy/wasp-dsl.md).
- The **TS Config** (`main.wasp.ts` with the class-based `new App(...)` API). See [Migrating from the TS Config](../guides/legacy/wasp-ts-config.md).
:::

## A quick example

```ts title="main.wasp.ts"
import { app, page, route, query } from '@wasp.sh/spec'
import MainPage from '@src/MainPage'

> import { getTasks } from '@src/queries'

Reviewer: I'd update these to ref imports as well.

- The **TS Config** (`main.wasp.ts` with the class-based `new App(...)` API). See [Migrating from the TS Config](../guides/legacy/wasp-ts-config.md).
:::

## A quick example

```ts title="main.wasp.ts"
import { app, page, route, query } from '@wasp.sh/spec'
import MainPage from '@src/MainPage'
import { getTasks } from '@src/queries'

export default app({
  name: 'todoApp',
  title: 'ToDo App',
  wasp: { version: '^0.24.0' },
  parts: [
    route('MainRoute', '/', page(MainPage, { authRequired: true })),
    query(getTasks, { entities: ['Task'] }),
  ],
})
```

You build your app by:

1. Importing the building blocks (`app`, `page`, `route`, `query`, ...) from `@wasp.sh/spec`.
2. Importing your own components and functions from `@src/...`.
3. Calling `app({ ... })` with your app's configuration, listing all the pages, routes, queries, actions, etc. in the `parts` array.
4. Exporting the result as the **default export** of the file.

## Pointing to your app's code

Anywhere the Wasp Spec expects your app's function or component (like a page's `component` or a query's `fn`), you can provide it in one of two ways:

> ### Magical imports {#magical-imports}

Reviewer: I think we can do better than "magical". From user's perspective they are "just regular JS imports" that require using a certain the `@src` alias. I know it's important for us to establish that these imports are "special" but maybe we can go with something like "`@src` imports" or "source imports" or "ergonomic imports"? 

I'd like to brainstorm the name.

Author: setting up a call for it

Author: Agreed in meeting: Reference imports

**Recommended**

Import the value with the regular syntax, and pass it directly.

```ts
import MainPage from '@src/MainPage'
import { getTasks } from '@src/queries'

export default app({
  parts: [
    page(MainPage),
    query(getTasks)
  ]
})
```

:::warning Limitations

Magical imports work by analyzing the file before executing it. As such, they have the following limitations:

- They only work for static imports (the `import ... from ...` syntax), not dynamic imports (the `import(...)` function).
- The referenced file must start with `@src/`.

> - Imports that reference other `*.wasp.ts` files are skipped, and processed normally at the spec level.

Reviewer: ```suggestion
- Imports from other `*.wasp.ts` files are treated like regular imports. 
```

Something like this maybe? Btw. I need to implement this. I'll link to the PR here.

Reviewer: Here's a draft PR for it: https://github.com/wasp-lang/wasp/pull/4160

Author: Oh I thought it was already implemented. I'll improve the docs but leave the feature in and see what we do with it.

  title: 'ToDo App',
  wasp: { version: '^0.24.0' },
  parts: [
    route('MainRoute', '/', page(MainPage, { authRequired: true })),
    ...auth,
  ],
})
```

All spec files should have the `.wasp.ts` extension, so they are included in the `tsconfig.wasp.json` and type-checked.

### Detecting production mode

Wasp sets the `NODE_ENV` environment variable based on which command you use to run Wasp:

- `"development"` during `wasp start` (and some other commands that compile the project, like `wasp db migrate-dev`).
- `"production"` during `wasp build`.

Because the Wasp Spec is just TypeScript, you can read this variable to switch config values per environment:

```ts title="main.wasp.ts"
const isProd = process.env.NODE_ENV === 'production'

export default app({
  //...
  emailSender: {
    provider: isProd ? 'SMTP' : 'Dummy',
    defaultFrom: { email: 'hi@example.com' },
  }
})
```

## Reference

### `@wasp.sh/spec` API

#### `ExtImport`

An **`ExtImport`** is a reference to your app's code. It can be either a [magical import](#magical-imports) or an [import object](#import-object).

#### `app`

Creates the app spec. Export the result as the file's default export.

> 

Reviewer: Idea: what if we used a table and not bullets to render this a bit more organised e.g. name, type, description. Then we might have nested tables for nested object or not. Just brainstorming here.

Author: I actually think we should autogenerate this.

Author: Now this is part of the autogenerated API docs

> - `config: object` the app configuration:

Reviewer: Nitpick: I find it odd to see `object` when it's probably something like `App`.

Author: Isn't `App` the _resulting_ type from the call? I think `object` is fine here.

Author: Now this is part of the autogenerated API docs

  - `name: string` internal app name (no spaces).
  - `title: string` browser tab title.

>   - `wasp: { version: string }` the Wasp version range to use.

Reviewer: Nitpick: why not nested bullet points since it's a nested object.

Author: Now this is part of the autogenerated API docs

>   - `head?: string[]` extra tags injected into the HTML `<head>`.

Reviewer: Do we mention it needs to be JSX?

>   - `auth?: Auth` authentication configuration.

Reviewer: Looks like we are missing details for `Auth`, `Server` and `Client` objects?

Author: I didn't want to go to in-depth, the TypeScript objects and the Auth docs will guide the user, no?

Author: Now this is part of the autogenerated API docs

  - `server?: Server` server setup, middleware, and env validation.
  - `client?: Client` client root component, setup, base dir, and env validation.
  - `db?: Db` database seeds and Prisma setup.
  - `emailSender?: EmailSender` email provider and default sender.
  - `webSocket?: WebSocket` WebSocket function and options.
  - `parts: Part[]` the pages, routes, queries, actions, APIs, jobs, and CRUDs (created with the constructors below).

> #### `page`

Reviewer: Idea: show example usage e.g. `page(HomePage, { authRequired: true })` for each declaration.

Author: Now this is part of the autogenerated API docs (they have examples)

Defines a page (a React component rendered for a route).

- `component: ExtImport` the React component for the page.
- `config?: object`
  - `authRequired?: boolean` if `true`, only authenticated users can access the page.

#### `route`

Maps a URL path to a page.

- `name: string` unique route name.
- `path: string` the URL path (e.g. `/`, `/tasks/:id`).

> - `page: Page` the page object returned by `page(...)`.

Reviewer: I'd maybe link to the `#### page` anchor

Author: Done in the autogenerated API docs

- `config?: object`
  - `lazy?: boolean` lazy-load the page's component.

>   - `prerender?: boolean` prerender the page at build time.

Reviewer: What are the defaults here, maybe we show that as well in the reference. What do you think?

Author: Done in the autogenerated API

#### `query`

Defines a read-only operation.

- `fn: ExtImport` the query function.
- `config?: object`

>   - `entities?: string[]` entities the query operates on.

Reviewer: I'd mention these are entity names defined in Prisma, not just any strings.

Author: Done in the autogenerated API docs

  - `auth?: boolean` whether the query requires authentication.

#### `action`

Defines an operation that modifies data.

- `fn: ExtImport` the action function.
- `config?: object`
  - `entities?: string[]` entities the action operates on.
  - `auth?: boolean` whether the action requires authentication.

#### `api`

Defines a custom HTTP endpoint.

- `method: 'ALL' | 'GET' | 'POST' | 'PUT' | 'DELETE'` the HTTP method.
- `path: string` the endpoint path.
- `fn: ExtImport` the request handler.
- `config?: object`

>   - `middlewareConfigFn?: ExtImport` customizes the middleware for this endpoint.

Reviewer: Maybe it would be good to cross-link to parts of the docs where we explain middleware customization.

Author: Done in the autogenerated API docs

  - `entities?: string[]` entities the handler operates on.
  - `auth?: boolean` whether the endpoint requires authentication.

#### `apiNamespace`

Applies shared middleware to all APIs under a path prefix.

- `path: string` the path prefix the namespace applies to.
- `config: object`
  - `middlewareConfigFn: ExtImport` customizes the middleware for the namespace.

#### `job`

Defines a background job.

- `fn: ExtImport` the function the job performs.
- `config: object`
  - `executor: 'PgBoss'` the job executor.
  - `schedule?: { cron: string; args?: object; executorOptions?: { pgBoss: object } }` runs the job on a cron schedule.
  - `entities?: string[]` entities the job operates on.
  - `performExecutorOptions?: { pgBoss: object }` executor options applied when the job is performed.

#### `crud`

Generates queries and actions for an entity.

- `name: string` unique CRUD name.
- `entity: string` the entity to generate operations for.
- `operations: object` a map of `get` / `getAll` / `create` / `update` / `delete`, each an object:
  - `isPublic?: boolean` whether the operation skips authentication.
  - `overrideFn?: ExtImport` a custom implementation for the operation.

> You can find more details about each of these features in the corresponding documentation for the features.

Reviewer: Yes, let's link in each place to the relevant docs. This makes the reference much more dense and useful for humans and LLMs IMHO

Author: Done in the autogenerated API docs
